### PR TITLE
fix: normalize pointers on cache for active entities

### PR DESCRIPTION
**Pull request by aleortega**
## Description

It fixes how the content server stores and retrieves Entities by Ids or Pointers.

Reason of the issue: we had keys mixing uppercases and lowercases on the cache.

## Changes

- Now the entities pointers are saved and retrieved from the cache always using lowercases
- Add 2 tests to verify that the issue is solved
- Also this solves the following edge case while keeping the ids cache key case sensitive:
```
1. If you try to get an entity by Id, the client needs to send that Id always in lowercases
2. If the client at some point miss parsing the Id to lowercase and send it with at least one letter in uppercase, then the cache breaks
3. This fix prevents corrupting the cache when an invalid id is asked to the catalyst, meaning that every time the client sends a valid id the peer answers with the proper object, independently from prior calls
```

## Break-down of prior versions issues

- v5.1.10
    1. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:basefemale"] }
        2. **Response:** Ok
    2. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:BaseFemale"] }
        2. **Response:** Ok
    3. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:BaseFemale"] }
        2. **Response:** Empty
    4. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:basefemale"] }
        2. **Response:** Ok
    - **Conslusion:** Only lowercases are allowed to retrieve entities by pointers
- v5.1.12
    1. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:basefemale"] }
        2. **Response:** Ok
    2. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:BaseFemale"] }
        2. **Response:** Empty
    3. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:BaseFemale"] }
        2. **Response:** Empty
    4. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:basefemale"] }
        2. **Response:** Empty
    5. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:basefemale"] }
        2. **Response:** Empty
    - **Conslusion:** Once a wearable is retrieved with mix cases, even the all lowercases approach won't work afterwards
- v5.1.13
    1. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:basefemale"] }
        2. **Response:** Ok
    2. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:BaseFemale"] }
        2. **Response:** Ok
    3. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:BaseFemale"] }
        2. **Response:** Empty
    4. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:basefemale"] }
        2. **Response:** Ok
     - **Conslusion:** Only lowercases are allowed to retrieve entities by pointers
- This fix
    1. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:basefemale"] }
        2. **Response:** Ok
    2. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:BaseFemale"] }
        2. **Response:** Ok
    3. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:BaseFemale"] }
        2. **Response:** Ok
    4. Request POST /entities/active
        1. **Body:** { pointers: ["urn:decentraland:off-chain:base-avatars:basefemale"] }
        2. **Response:** Ok
    - **Conslusion:** Working as expected

### CODE REVIEW

File: content/src/ports/activeEntities.ts [hunk 1]
```diff
@@ -106,7 +106,7 @@ export const createActiveEntitiesComponent = (
   const update = async (pointers: string[], entity: Entity | NotActiveEntity): Promise<void> => {
     for (const pointer of pointers) {
       setPreviousEntityAsNone(pointer)
-      entityIdByPointers.set(pointer, isEntityPresent(entity) ? entity.id : entity)
+      entityIdByPointers.set(normalizePointer(pointer), isEntityPresent(entity) ? entity.id : entity)
```
Comment (agusaldasoro) [inline content/src/ports/activeEntities.ts:109]: should we do something similar for the queries by entityId and URN prefix?

Comment (aleortega) [reply content/src/ports/activeEntities.ts:109]: I'm on it, thanks! Will test the id workflow as well.

Comment (aleortega) [reply content/src/ports/activeEntities.ts]: I updated the approach to also cover the edge cases related to the ids. Anyways I still have doubts about the entities id management in the database, if the database is case insensitive by always using lowercases this solution is safe, otherwise I think this won't work as expected and we will have further bugs.

Comment (aleortega) [reply content/src/ports/activeEntities.ts]: These changes about normalizing also the ids fixed the corner case explained in this PR description -> https://github.com/decentraland/catalyst/pull/1380

**Comment by pedrotambo** [comment]
I wonder if a wrapper for the cache would make sense to avoid future bugs of that kind, such us forgetting to use `.toLowerCase`.
What about something like this?
```function createEntityByPointersCache(): Map<string, string | NotActiveEntity> {
  const entityIdByPointers = new Map<string, string | NotActiveEntity>()
  return {
    ...entityIdByPointers,
    get(pointer: string) {
      return entityIdByPointers.get(pointer.toLowerCase())
    },
    set(pointer: string, entity: string | NotActiveEntity) {
      return entityIdByPointers.set(pointer.toLocaleLowerCase(), entity)
    }
  }
}
```

File: content/src/ports/activeEntities.ts (hunk 1 above)
Comment (aleortega) [reply content/src/ports/activeEntities.ts]: After confirming that the entity ids should be case sensitive we reverted this change. Anyways, we took advantage of these modifications to solve the corner case explained in the PR description.

**Comment by aleortega** [comment]
> I wonder if a wrapper for the cache would make sense to avoid future bugs of that kind, such us forgetting to use `.toLowerCase`. What about something like this?
> 
> ```
> function createEntityByPointersCache(): Map<string, string | NotActiveEntity> {
>   const entityIdByPointers = new Map<string, string | NotActiveEntity>()
>   return {
>     ...entityIdByPointers,
>     get(pointer: string) {
>       return entityIdByPointers.get(pointer.toLowerCase())
>     },
>     set(pointer: string, entity: string | NotActiveEntity) {
>       return entityIdByPointers.set(pointer.toLocaleLowerCase(), entity)
>     }
>   }
> }
> ```

I like this, let me try to implement it, thank you!

**Review by pedrotambo — APPROVED** [review]
LGTM

**Comment by menduz** [comment]
Adding extra test endpoints

- https://peer-ue-2.decentraland.zone/content/entities/wearables/?pointer=urn:decentraland:off-chain:base-avatars:basefemale
- https://peer-ue-2.decentraland.zone/content/entities/wearables/?pointer=urn:decentraland:off-chain:base-avatars:BaseFemale
- https://peer-ue-2.decentraland.zone/lambdas/collections/wearables/?wearableId=urn:decentraland:off-chain:base-avatars:basefemale
- https://peer-ue-2.decentraland.zone/lambdas/collections/wearables/?wearableId=urn:decentraland:off-chain:base-avatars:BaseFemale

**Comment by aleortega** [comment]
> Adding extra test endpoints
> 
> * https://peer-ue-2.decentraland.zone/content/entities/wearables/?pointer=urn:decentraland:off-chain:base-avatars:basefemale
> * https://peer-ue-2.decentraland.zone/content/entities/wearables/?pointer=urn:decentraland:off-chain:base-avatars:BaseFemale
> * https://peer-ue-2.decentraland.zone/lambdas/collections/wearables/?wearableId=urn:decentraland:off-chain:base-avatars:basefemale
> * https://peer-ue-2.decentraland.zone/lambdas/collections/wearables/?wearableId=urn:decentraland:off-chain:base-avatars:BaseFemale

I'm on it, will post the results of each endpoint with the final version in here.

**Comment by aleortega** [comment]
> > Adding extra test endpoints
> > 
> > * https://peer-ue-2.decentraland.zone/content/entities/wearables/?pointer=urn:decentraland:off-chain:base-avatars:basefemale
> > * https://peer-ue-2.decentraland.zone/content/entities/wearables/?pointer=urn:decentraland:off-chain:base-avatars:BaseFemale
> > * https://peer-ue-2.decentraland.zone/lambdas/collections/wearables/?wearableId=urn:decentraland:off-chain:base-avatars:basefemale
> > * https://peer-ue-2.decentraland.zone/lambdas/collections/wearables/?wearableId=urn:decentraland:off-chain:base-avatars:BaseFemale
> 
> I'm on it, will post the results of each endpoint with the final version in here.

Just executed the manual tests, not sure if we also need Integration Tests for these endpoints since I think the automated tests added in this PR already tests these cases because they use the same service, but please let me know if I'm wrong:

- Fix on lambdas/collections/wearables
    1. Request GET lambdas/collections/wearables
        1. **Query String:** ?wearableId=urn:decentraland:off-chain:base-avatars:basefemale
        2. **Response:** Ok
    2. Request GET lambdas/collections/wearables
        1. **Query String:** ?wearableId=urn:decentraland:off-chain:base-avatars:BaseFemale
        2. **Response:** Ok
    3. Request GET lambdas/collections/wearables
        1. **Query String:** ?wearableId=urn:decentraland:off-chain:base-avatars:BaseFemale
        2. **Response:** Ok
    4. Request GET lambdas/collections/wearables
        1. **Query String:** ?wearableId=urn:decentraland:off-chain:base-avatars:basefemale
        2. **Response:** Ok
    - **Conslusion:** Working as expected
- Fix on content/entities/wearables
    1. Request GET content/entities/wearables
        1. **Query String:** ?pointer=urn:decentraland:off-chain:base-avatars:basefemale
        2. **Response:** Ok
    2. Request GET content/entities/wearables
        1. **Query String:** ?pointer=urn:decentraland:off-chain:base-avatars:BaseFemale
        2. **Response:** Ok
    3. Request GET content/entities/wearables
        1. **Query String:** ?pointer=urn:decentraland:off-chain:base-avatars:BaseFemale
        2. **Response:** Ok
    4. Request GET content/entities/wearables
        1. **Query String:** ?pointer=urn:decentraland:off-chain:base-avatars:basefemale
        2. **Response:** Ok
    - **Conslusion:** Working as expected

cc @menduz

**Comment by aleortega** [comment]
> > > Adding extra test endpoints
> > > 
> > > * https://peer-ue-2.decentraland.zone/content/entities/wearables/?pointer=urn:decentraland:off-chain:base-avatars:basefemale
> > > * https://peer-ue-2.decentraland.zone/content/entities/wearables/?pointer=urn:decentraland:off-chain:base-avatars:BaseFemale
> > > * https://peer-ue-2.decentraland.zone/lambdas/collections/wearables/?wearableId=urn:decentraland:off-chain:base-avatars:basefemale
> > > * https://peer-ue-2.decentraland.zone/lambdas/collections/wearables/?wearableId=urn:decentraland:off-chain:base-avatars:BaseFemale
> > 
> > 
> > I'm on it, will post the results of each endpoint with the final version in here.
> 
> Just executed the manual tests, not sure if we also need Integration Tests for these endpoints since I think the automated tests added in this PR already tests these cases because they use the same service, but please let me know if I'm wrong:
> 
> * Fix on lambdas/collections/wearables
>   
>   1. Request GET lambdas/collections/wearables
>      
>      1. **Query String:** ?wearableId=urn:decentraland:off-chain:base-avatars:basefemale
>      2. **Response:** Ok
>   2. Request GET lambdas/collections/wearables
>      
>      1. **Query String:** ?wearableId=urn:decentraland:off-chain:base-avatars:BaseFemale
>      2. **Response:** Ok
>   3. Request GET lambdas/collections/wearables
>      
>      1. **Query String:** ?wearableId=urn:decentraland:off-chain:base-avatars:BaseFemale
>      2. **Response:** Ok
>   4. Request GET lambdas/collections/wearables
>      
>      1. **Query String:** ?wearableId=urn:decentraland:off-chain:base-avatars:basefemale
>      2. **Response:** Ok
>   
>   
>   * **Conslusion:** Working as expected
> * Fix on content/entities/wearables
>   
>   1. Request GET content/entities/wearables
>      
>      1. **Query String:** ?pointer=urn:decentraland:off-chain:base-avatars:basefemale
>      2. **Response:** Ok
>   2. Request GET content/entities/wearables
>      
>      1. **Query String:** ?pointer=urn:decentraland:off-chain:base-avatars:BaseFemale
>      2. **Response:** Ok
>   3. Request GET content/entities/wearables
>      
>      1. **Query String:** ?pointer=urn:decentraland:off-chain:base-avatars:BaseFemale
>      2. **Response:** Ok
>   4. Request GET content/entities/wearables
>      
>      1. **Query String:** ?pointer=urn:decentraland:off-chain:base-avatars:basefemale
>      2. **Response:** Ok
>   
>   
>   * **Conslusion:** Working as expected
> 
> cc @menduz

Since all the tests passed (manual and automated) I'll merge this PR, anyways if we want to add those Integration Tests commented above I'll take ownership of them and raise a new PR for it. The main reason of closing this PR is to avoid delaying one more day the `Solution Completed` field of the `RCA` which can be found here (I still need to update that day to place today instead of yesterday) -> https://github.com/decentraland/rca/blob/main/incidents/2022-12-27.md
